pkg/operator/controller/dnsnameresolver-feature/controller.go

```diff
@@ -0,0 +1,160 @@
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	controllerName = "dnsnameresolverfeature_controller"
+)
+
+// New creates and returns a controller that creates DNSNameResolver CRD when the
+// appropriate featuregate is enabled.
+func New(mgr manager.Manager, config Config) (controller.Controller, error) {
+	reconciler := &reconciler{
+		cache:  mgr.GetCache(),
+		client: mgr.GetClient(),
+		config: config,
+	}
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+	clusterNamePredicate := predicate.NewPredicateFuncs(func(o client.Object) bool {
+		expectedName := operatorcontroller.FeatureGateClusterConfigName()
+		actualName := types.NamespacedName{
+			Namespace: o.GetNamespace(),
+			Name:      o.GetName(),
+		}
+		return expectedName == actualName
+	})
+	if err := c.Watch(source.Kind(mgr.GetCache(), &configv1.FeatureGate{}), &handler.EnqueueRequestForObject{}, clusterNamePredicate); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// Config holds all the configuration that must be provided when creating the
+// controller.
+type Config struct {
+	// DNSNameResolverEnabled indicates that the "DNSNameResolver" featuregate is enabled.
+	DNSNameResolverEnabled bool
+
+	FeatureGateName string
+
+	// DependentCaches is a list of caches that are used by Controllers watching DNSNameResolver
+	// resources. The dnsnameresolverfeature controller starts these caches once
+	// the DNSNameResolver CRD has been created.
+	DependentCaches []cache.Cache
+	// DependentControllers is a list of controllers that watch DNSNameResolver
+	// resources. The dnsnameresolverfeature controller starts these controllers once
+	// the DNSNameResolver CRD has been created and the DependentCaches are started.
+	DependentControllers []controller.Controller
+}
+
+// reconciler handles the actual featuregate reconciliation logic in response to
+// events.
+type reconciler struct {
+	config Config
+
+	cache            cache.Cache
+	client           client.Client
+	startCaches      sync.Once
+	startControllers sync.Once
+}
+
+// Reconcile expects request to refer to a FeatureGate and creates or
+// reconciles the DNSNameResolver CRD.
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	logrus.Info("reconciling", "request", request)
+
+	// Check if the DNSNameResolved feature gate is enabled or not.
+	if !r.config.DNSNameResolverEnabled {
+		return reconcile.Result{}, nil
+	}
+
+	// Check if the name of feature gate matches with the default feature gate.
+	if request.Name != r.config.FeatureGateName {
+		// Return a nil error value to avoid re-triggering the event.
+		logrus.Errorf("skipping unexpected feature gate object %s", request.Name)
```

Comment:
If this is important enough to be logged as an error then maybe this condition should return an error?

Reply:
Actually this check is not required as predicate added to the Watch of FeatureGate objects will only reconcile if the name of the object matches the default feature gate name. 
https://github.com/openshift/cluster-dns-operator/pull/394/files#diff-39dda6125e447d50ba6c620c266c0d064116e52b6c7c4e2e12f4783c15f2d7d6R40-R50

I'll remove this if statement.